src/odemis/acq/move.py

```diff
@@ -614,6 +618,437 @@ def _doCryoSwitchSamplePosition(self, future, target):
+                       [max_x, min_y],
+                       [max_x, max_y]]
+
+        # Calculate range based on the conversion
+        new_coordinates = list(map(self._convertPosFromdep, coordinates))
+
+        # Update default axes range
+        new_range = numpy.array(new_coordinates, dtype=float)
+        self.axes["x"].range = (min(new_range[:, 0]), max(new_range[:, 0]))
+        self.axes["y"].range = (min(new_range[:, 1]), max(new_range[:, 1]))
+
+
+class MeteorTFS2PostureManager(MeteorTFS1PostureManager):
+    def __init__(self, microscope):
+        super().__init__(microscope)
+        # Check required metadata used during switching
+        self.required_keys.add(model.MD_POS_COR)
+        self.check_stage_metadata(required_keys=self.required_keys)
+        if not {"x", "y", "rz", "rx"}.issubset(self.stage.axes):
+            raise KeyError("The stage misses 'x', 'y', 'rx' or 'rz' axes")
+        stage_md = self.stage.getMetadata()
+        self.rotation = stage_md[model.MD_ROTATION_COR]  # pre-tilt of the shuttle
+        # Instantiate a class to convert stage axes such that the imaging plane in the new axes
+        # moves perpendicular to the objective axis by keeping a fixed distance
+        self.convert_stage = ConvertStage(self.stage, ["y", "z"], self.rotation)
+        # Helpful parameters to work with fixed imaging plane in FM
+        self.stage_pos_before_fav_fm = None  # previous stage position in SEM before switching to FM
+        self.fav_pos_active = None  # desired imaging plane axes values
+        self.update_fav_pos_active()  # estimate and update the desired imaging plane axes values
+
+    def update_fav_pos_active(self):
+        """
+        In FM mode, observe the features on sample stage at a preferred plane defined by fav_pos_active
+        such that the sample plan in FM moves in a defined 2D plane
+        """
+        # calculate favorite position from the centre of grid 2
+        stage_md = self.stage.getMetadata()
+        sem_grid2_pos = stage_md[model.MD_SAMPLE_CENTERS][POSITION_NAMES[GRID_2]]
+        sem_grid2_pos.update(stage_md[model.MD_FAV_SEM_POS_ACTIVE])
+        end_pos = self.calculate_pos_before_fixing_imaging_plane(sem_grid2_pos)
+        self.fav_pos_active = {"z": (end_pos["y"] * math.sin(self.rotation) +
+                                     end_pos["z"] * math.cos(self.rotation))}
+
+    # Note: this transformation consists of translation of along x and y
+    # axes, and 7 degrees rotation around rx, and 180 degree rotation around rz.
+    # The rotation angles are constant existing in "FM_POS_ACTIVE" metadata,
+    # but the translation are calculated based on the current position and some
+    # correction/shifting parameters existing in metadata "FM_POS_ACTIVE".
+    # This correction parameters can change every session. They are calibrated
+    # at the beginning of each run.
+    def _transformFromSEMToMeteor(self, pos: Dict[str, float]) -> Dict[str, float]:
```

Comment:
Instead of having two functions that do very similar thing (except one condition), merge _transformFromSEMToMeteor and calculate_pos_before_fixing_imaging_plane. Just pass an argument when you want to fix the imaging plane